### PR TITLE
Add display name after create context

### LINKED ISSUE #11241: Support setting Context displayName in `@babel/plugin-transform-react-display-name`
## Feature Request

- [x] I would like to work on this feature!
- [ ] I have bandwidth to work on this feature in the near future. 🙁

**Is your feature request related to a problem? Please describe.**

Here's an example of React DevTools with a few context providers and consumers:

image

You'll notice that the context consumer and provider rendered under `ThemeProvider` have generic `Context.Consumer` and `Context.Provider` display names, but there's also a `Router.Provider` and `Router.Consumer` in there which are much more clear.

They accomplish this by setting the context's `displayName` value.

Here are some test cases:

```javascript
import React from 'react'

const MyContext = React.createContext()
MyContext.displayName = 'MyContext'
```

**Describe the solution you'd like**

I'd like [`@babel/plugin-transform-react-display-name`](https://github.com/babel/babel/blob/master/packages/babel-plugin-transform-react-display-name/src/index.js) to support automatically setting the `displayName` property for contexts.

**Describe alternatives you've considered**

I can set it manually. Not desirable. Or I could write my own babel plugin, but then others wouldn't benefit from it if they either don't know about the plugin or can't install it. Also, this seems to make a lot of sense to be included here to me.

**Teachability, Documentation, Adoption, Migration Strategy**

No need, it'll just magically start working and people will be pleasantly surprised.

### PR DESCRIPTION
**Pull request by JLHwung**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #11241 (28👍)
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      | Yes
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

The PR inserts `context.displayName = "context"` assignments after `React.createContext()` call. For common cases like variable declarations and assignment, we reuse the context binding when building assignments. Then we fallback to the sequence expression approach, where `React.createContext()` is transformed into `(ref = React.createContext(), ref.displayName = "myContext", ref)`.

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/13501"></a>

### CODE REVIEW

File: packages/babel-plugin-transform-react-display-name/src/index.js [hunk 1]
```diff
@@ -2,29 +2,111 @@ import { declare } from "@babel/helper-plugin-utils";
...
+
+  // ensure that we have an identifier we can inherit from
+  if (!id) return;
+
+  // foo.bar -> bar
+  if (t.isMemberExpression(id)) {
+    id = id.property;
   }
 
+  return id;
```
Comment (nicolo-ribaudo) [inline packages/babel-plugin-transform-react-display-name/src/index.js:55]: Here `id` could be any expression. Given the function name, maybe it makes sense to return `undefined` if it's not an identifier?

File: packages/babel-plugin-transform-react-display-name/src/index.js [hunk 2]
```diff
@@ -2,29 +2,111 @@ import { declare } from "@babel/helper-plugin-utils";
...
+}
+
+function isCreateContext(node) {
+  let callee;
+  return (
+    t.isCallExpression(node) &&
+    t.isMemberExpression((callee = node.callee)) &&
+    t.isIdentifier(callee.object, { name: "React" }) &&
+    (t.isIdentifier(callee.property, { name: "createContext" }) ||
```
Comment (nicolo-ribaudo) [inline packages/babel-plugin-transform-react-display-name/src/index.js:64]: This would also allow `React[createContext]`.

**Review by nicolo-ribaudo — COMMENTED** [review]
This should probably go in a minor?

File: packages/babel-plugin-transform-react-display-name/src/index.js (hunk 1 above)
Comment (JLHwung) [reply packages/babel-plugin-transform-react-display-name/src/index.js:55]: We currently check `t.isIdentifier` after `getDisplayNameReferenceIdentifier`. I can move the check here.

**Review by existentialism — APPROVED** [review]

**Review by nicolo-ribaudo — APPROVED** [review]

File: packages/babel-plugin-transform-react-display-name/src/index.js [hunk 3]
```diff
@@ -2,28 +2,116 @@ import { declare } from "@babel/helper-plugin-utils";
...
+  if (parentPath.isVariableDeclarator()) {
+    // FooContext = React.createContext()
+    const ref = parentPath.node.id;
+    // parentPath.parentPath must be a VariableDeclaration because getDisplayNameReferenceIdentifier
+    // does not support patterns
+    parentPath.parentPath.insertAfter(buildDisplayNameAssignment(ref, id));
+  } else if (parentPath.isAssignmentExpression()) {
+    // var FooContext = React.createContext()
+    const ref = parentPath.node.left;
+    parentPath.insertAfter(buildDisplayNameAssignment(ref, id));
```
Comment (JLHwung) [inline packages/babel-plugin-transform-react-display-name/src/index.js:95]: Inserting after an assignment expression does not preserve completion records. Maybe we should only insert for variable declarations.

Comment (nicolo-ribaudo) [reply packages/babel-plugin-transform-react-display-name/src/index.js:95]: Inserting after a variable declaration changes the completion record too.

This is a general problem with `insertAfter` and it's not just in this PR. I think we should check it separately checking in `insertAfter` if the completion record matters (e.g. if a parent is a `do` expression) and injeting a temp variable similarly to what we do for `insertAfter` in expressions.

Comment (JLHwung) [reply packages/babel-plugin-transform-react-display-name/src/index.js:95]: The current version of do expression forbids a do block ending with variable declarations: https://tc39.es/proposal-do-expressions/#sec-doexpression-static-semantics-early-errors, but we haven't implemented that in Babel parser. Guess it should be okay for variable declarations (most common cases) here?

Comment (nicolo-ribaudo) [reply packages/babel-plugin-transform-react-display-name/src/index.js:95]: Uh well ok then, I guess assignment expressions in this case are not common.

**Comment by csestito** [comment]
Hi @JLHwung . I am having issues in a [different project](https://github.com/facebook/docusaurus/issues/5286) today and it seems related to this change. The error we are getting: 

```./node_modules/@docusaurus/theme-classic/lib-next/theme/ThemeContext.js 6:130
Module parse failed: Unexpected token (6:130)
File was processed with these loaders:
 * ./node_modules/cache-loader/dist/cjs.js
 * ./node_modules/@docusaurus/core/node_modules/babel-loader/lib/index.js
You may need an additional loader to handle the result of these loaders.
|  * This source code is licensed under the MIT license found in the
|  * LICENSE file in the root directory of this source tree.
>  */import React from'react';const ThemeContext=/*#__PURE__*/React.createContext(undefined);ThemeContext.displayName="ThemeContext"export default ThemeContext;
 @ ./node_modules/@docusaurus/theme-classic/lib-next/theme/hooks/useThemeContext.js 6:33-79 6:131-143
 @ ./node_modules/@docusaurus/theme-classic/lib-next/theme/Navbar/index.js
 @ ./node_modules/@docusaurus/theme-classic/lib-next/theme/Layout/index.js
 @ ./node_modules/@docusaurus/theme-classic/lib-next/theme/NotFound.js
 @ ./node_modules/@docusaurus/core/lib/client/exports/ComponentCreator.js
 @ ./.docusaurus/routes.js
 @ ./node_modules/@docusaurus/core/lib/client/clientEntry.js
```
 
The missing `;` between `ThemeContext.displayName="ThemeContext"` and `export default ThemeContext;` is causing parser errors. Is this something that needs fixed elsewhere, or should it be part of the plugin code?

Thanks!

**Pull request merged**